File: app/src/main/java/wangdaye/com/geometricweather/background/polling/work/worker/TodayForecastUpdateWorker.java
package wangdaye.com.geometricweather.background.polling.work.worker;

import android.annotation.SuppressLint;
import android.content.Context;

import androidx.annotation.NonNull;
import androidx.work.WorkerParameters;
import androidx.work.impl.utils.futures.SettableFuture;

import java.util.List;

import wangdaye.com.geometricweather.background.polling.PollingManager;
import wangdaye.com.geometricweather.basic.model.Location;
import wangdaye.com.geometricweather.remoteviews.presenter.notification.ForecastNotificationIMP;

public class TodayForecastUpdateWorker extends AsyncUpdateWorker {

    public TodayForecastUpdateWorker(@NonNull Context context, @NonNull WorkerParameters workerParams) {
        super(context, workerParams);
    }

    @Override
    public void updateView(Context context, Location location) {
        if (ForecastNotificationIMP.isEnable(context, true)) {
            ForecastNotificationIMP.buildForecastAndSendIt(context, location, true);
        }
    }

    @Override
    public void updateView(Context context, List<Location> locationList) {
    }

    @SuppressLint("RestrictedApi")
    @Override
    public void handleUpdateResult(SettableFuture<Result> future, boolean failed) {
        future.set(failed ? Result.failure() : Result.success());
        PollingManager.resetTodayForecastBackgroundTask(
                getApplicationContext(), false, true);
    }
}

File: app/src/main/java/org/breezyweather/background/polling/work/worker/TodayForecastUpdateWorker.kt
package org.breezyweather.background.polling.work.worker

import android.annotation.SuppressLint
import android.content.Context
import androidx.hilt.work.HiltWorker
import androidx.work.WorkerParameters
import androidx.work.impl.utils.futures.SettableFuture
import dagger.assisted.Assisted
import dagger.assisted.AssistedInject
import org.breezyweather.background.polling.PollingManager.resetTodayForecastBackgroundTask
import org.breezyweather.common.basic.models.Location
import org.breezyweather.location.LocationHelper
import org.breezyweather.remoteviews.presenters.notification.ForecastNotificationIMP.buildForecastAndSendIt
import org.breezyweather.remoteviews.presenters.notification.ForecastNotificationIMP.isEnabled
import org.breezyweather.weather.WeatherHelper

@HiltWorker
class TodayForecastUpdateWorker @AssistedInject constructor(
    @Assisted context: Context,
    @Assisted workerParams: WorkerParameters,
    locationHelper: LocationHelper,
    weatherHelper: WeatherHelper
) : AsyncUpdateWorker(context, workerParams, locationHelper, weatherHelper) {
    override fun updateView(context: Context, location: Location) {
        if (isEnabled(context, true)) {
            buildForecastAndSendIt(context, location, true)
        }
    }

    override fun updateView(context: Context, locationList: List<Location>) {}

    @SuppressLint("RestrictedApi")
    override fun handleUpdateResult(future: SettableFuture<Result>?, failed: Boolean) {
        future?.set(if (failed) Result.failure() else Result.success())
        resetTodayForecastBackgroundTask(
            applicationContext,
            forceRefresh = false,
            nextDay = true
        )
    }
}
